src/pmdas/bpf/modules/bashreadline.c

```diff
@@ -0,0 +1,395 @@
+/*
+ *
+ * Copyright (c) 2022 Sohaib Mohamed <sohaib.amhmd@gmail.com>
+ *
+ * Based on the bashreadline(8):
+ * https://github.com/iovisor/bcc/blob/master/libbpf-tools/bashreadline.c
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation; either version 2 of the License, or (at your
+ * option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * for more details.
+ */
+
+#include "module.h"
+
+#include <bpf/bpf.h>
+#include <bpf/libbpf.h>
+#include <pcp/pmda.h>
+#include <pcp/pmwebapi.h>
+#include <sys/queue.h>
+
+#include "bashreadline.h"
+#include "bashreadline.skel.h"
+#include "btf_helpers.h"
+#include "trace_helpers.h"
+#include "uprobe_helpers.h"
+
+#define PERF_BUFFER_PAGES 16
+#define PERF_POLL_TIMEOUT_MS 100
+#define INDOM_COUNT 2
+
+static struct env {
+    int process_count;
+} env = {
+    .process_count = 20,
+};
+
+static char *libreadline_path = NULL;
+
+static pmdaInstid *bashreadline_instances, *bashreadline_lost;
+static struct bashreadline_bpf *obj;
+static struct perf_buffer *pb = NULL;
+static char *readline_so_path;
+static struct str_t *str_t;
+static int lost_events;
+static int queuelength;
+
+/* cache array */
+struct tailq_entry {
+    struct str_t str_t;
+    TAILQ_ENTRY(tailq_entry) entries;
+};
+
+TAILQ_HEAD(tailhead, tailq_entry) head;
+
+static struct tailq_entry* allocElm(void)
+{
+    return malloc(sizeof(struct tailq_entry));
+}
+
+static void push(struct tailq_entry *elm)
+{
+    if (queuelength > env.process_count)
+    {
+        struct tailq_entry *l;
+        l = head.tqh_first;
+        TAILQ_REMOVE(&head, l, entries);
+        free(l);
+        queuelength--;
+    }
+    TAILQ_INSERT_TAIL(&head, elm, entries);
+    queuelength++;
+}
+
+static bool get_item(unsigned int offset, struct tailq_entry** val)
+{
+    struct tailq_entry *i;
+    unsigned int iter = 0;
+
+    TAILQ_FOREACH_REVERSE(i, &head, tailhead, entries) {
+        if (offset == iter) {
+            *val = i;
+            return true;
+        }
+        iter++;
+    }
+    return false;
+}
+
+static unsigned int indom_id_mapping[INDOM_COUNT];
+
+#define METRIC_COUNT 3
+enum metric_name { COMM, PID, LOST };
+enum metric_indom { BASHREADLINE_INDOM, LOST_EVENTS };
+
+char* metric_names[METRIC_COUNT] = {
+    [COMM] = "bashreadline.comm",
+    [PID]  = "bashreadline.pid",
+    [LOST] = "bashreadline.lost",
+};
+
+char* metric_text_oneline[METRIC_COUNT] = {
+    [COMM] = "Command name",
```

Comment:
Done. but I'm not sure if I did it the right way.

Reply:
I added LCFLAGS at the beginning of modules/GNUMakefile, which breaks trace_helpers.c!
so I want to enable std=c99 for all modules, expect *_helpers.c files...

Reply:
Hm, that's weird. Apparently:

* `typeof` is not part of the C99 standard (it will be a standard in C23), but it's available as a GNU extension (clang supports it as well)
* when we use `--std=c99` we disable GNU extensions [1] (therefore `typeof` doesn't work)
* when we don't use `--std=c99`, the gcc on my system (version 12.1.1) defaults to `--std=gnu17` [2], i.e. C17 with GNU extensions (therefore `typeof` works)
* however, we can use `__typeof__` instead of `typeof` and then it also works with `--std=c99`

So the irony is that by using `--std=c99` on a modern compiler we effectively downgrade the used C standard version. To fix this issue, I suggest to just change `typeof` to `__typeof__` in `trace_helpers.c`, then it compiles fine when using `LCFLAGS = $(C99_CFLAGS)` in the `GNUmakefile`.

[1] https://stackoverflow.com/a/17156740
[2] you can check this with `info gcc` --> `Standards` which reads `The default, if no C language dialect options are given, is
'-std=gnu17'.` on my system.

Reply:
Thanks for your time and this informative comment,
Since we are using `trace_helpers.c` from symlink, maybe It's not a good idea to patch that file? especially since this file changes quickly...

Reply:
afaics git automatically merges it, so unless there is a change in exactly that line, it shouldn't be an issue to modify this file from the remote repo.